b2/transferer.py

```diff
@@ -145,3 +230,79 @@ def as_info_dict(self):
             'contentSha1': self.content_sha1,
             'fileInfo': self.file_info,
         }
+
+
+class WriterThread(threading.Thread):
+    def __init__(self, file):
+        self.file = file
+        self.queue = queue.Queue()
+        super(WriterThread, self).__init__()
+
+    def run(self):
+        file = self.file
+        queue_get = self.queue.get
+        while 1:
+            shutdown, offset, data = queue_get()
+            if shutdown:
+                break
+            file.seek(offset)
+            file.write(data)
+
+    def __enter__(self):
+        self.start()
+        return self
+
+    def __exit__(self, exc_type, exc_val, exc_tb):
+        self.queue.put((True, None, None))
+        self.join()
+
+
+class HashingDownloaderThread(threading.Thread):
```

Comment:
This class differs in two ways from the NonHashing version: (1) it hashes, and (2) it stops reading before the end of the response body.  I think that's worth at least a comment, and maybe a different class name.  Maybe just call it `FirstPartDownloaderThread`